kotlin-reference-server/src/main/kotlin/org/stellar/reference/ReferenceServer.kt

```diff
@@ -1,73 +1,36 @@
 package org.stellar.reference
 
 import com.sksamuel.hoplite.*
-import io.ktor.http.*
-import io.ktor.serialization.kotlinx.json.*
-import io.ktor.server.application.*
-import io.ktor.server.auth.*
-import io.ktor.server.auth.jwt.*
-import io.ktor.server.engine.*
 import io.ktor.server.netty.*
-import io.ktor.server.plugins.contentnegotiation.*
-import io.ktor.server.plugins.cors.routing.*
-import io.ktor.server.response.*
 import mu.KotlinLogging
-import org.stellar.reference.data.Config
-import org.stellar.reference.data.LocationConfig
-import org.stellar.reference.plugins.*
+import org.stellar.reference.di.ConfigContainer
+import org.stellar.reference.di.EventConsumerContainer
+import org.stellar.reference.di.ReferenceServerContainer
 
 val log = KotlinLogging.logger {}
-lateinit var referenceKotlinSever: NettyApplicationEngine
+lateinit var referenceKotlinServer: NettyApplicationEngine
 
 fun main(args: Array<String>) {
   startServer(null, args.getOrNull(0)?.toBooleanStrictOrNull() ?: true)
 }
 
 fun startServer(envMap: Map<String, String>?, wait: Boolean) {
-  log.info { "Starting Kotlin reference server" }
-
   // read config
-  val cfg = readCfg(envMap)
-
-  // start server
-  referenceKotlinSever =
-    embeddedServer(Netty, port = cfg.appSettings.port) {
-        install(ContentNegotiation) { json() }
-        configureAuth(cfg)
-        configureRouting(cfg)
-        install(CORS) {
-          anyHost()
-          allowHeader(HttpHeaders.Authorization)
-          allowHeader(HttpHeaders.ContentType)
-        }
-        install(RequestLoggerPlugin)
-        install(RequestExceptionHandlerPlugin)
-      }
-      .start(wait)
-}
-
-fun readCfg(envMap: Map<String, String>?): Config {
-  // Load location config
-  val locationCfg =
-    ConfigLoaderBuilder.default()
-      .addPropertySource(PropertySource.environment())
-      .build()
-      .loadConfig<LocationConfig>()
+  ConfigContainer.init(envMap)
 
-  val cfgBuilder = ConfigLoaderBuilder.default()
-  // Add environment variables as a property source.
-  cfgBuilder.addPropertySource(PropertySource.environment())
-  envMap?.run { cfgBuilder.addMapSource(this) }
-  // Add config file as a property source if valid
-  locationCfg.fold({}, { cfgBuilder.addFileSource(it.ktReferenceServerConfig) })
-  // Add default config file as a property source.
-  cfgBuilder.addResourceSource("/default-config.yaml")
+  Thread {
```

Comment:
Start a coroutine instead?

Reply:
Hm.... the `KafkaConsumer` API is blocking so is coroutine the right choice here?

Reply:
If we remove it it doesn't matter :)
But in general you just dispatch a coroutine and it can work in the background. Whatever happens in coroutine can be blocking